frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BasePullBranchAction.java

```diff
@@ -0,0 +1,87 @@
+import com.virtuslab.logger.IPrefixedLambdaLogger;
+import com.virtuslab.logger.PrefixedLambdaLoggerFactory;
+
+/**
+ * Expects DataKeys:
+ * <ul>
+ *  <li>{@link DataKeys#KEY_GIT_MACHETE_REPOSITORY}</li>
+ *  <li>{@link CommonDataKeys#PROJECT}</li>
+ * </ul>
+ */
+public abstract class BasePullBranchAction extends GitMacheteRepositoryReadyAction {
+  private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("frontendActions");
+
+  protected final List<SyncToRemoteStatus.Relation> PULL_ENABLING_STATUSES = List.of(
+      SyncToRemoteStatus.Relation.Behind);
+
+  /**
+   * Bear in mind that {@link AnAction#beforeActionPerformedUpdate} is called before each action.
+   * (For more details check {@link com.intellij.openapi.actionSystem.ex.ActionUtil} as well.)
+   * The {@link AnActionEvent} argument passed to before-called {@link AnAction#update} is the same one that is passed here.
+   * This gives us certainty that all checks from actions' update implementations will be performed
+   * and all data available via data keys in those {@code update} implementations will still do be available
+   * in {@link BasePullBranchAction#actionPerformed} implementations.
+   */
+  @Override
+  @UIEffect
+  public abstract void actionPerformed(AnActionEvent anActionEvent);
+
+  protected void doPull(Project project, GitRepository gitRepository, String branchName) {
+    var trackingInfo = gitRepository.getBranchTrackInfo(branchName);
+
+    if (trackingInfo == null) {
+      LOG.warn("No branch tracking info for branch ${branchName}");
+      return;
+    }
+
+    new Task.Backgroundable(project, "Updating...", true) {
+
+      @Override
+      public void run(ProgressIndicator indicator) {
+        var fetchSupport = GitFetchSupportImpl.fetchSupport(project);
+        var fetchResult = fetchSupport.fetch(gitRepository, trackingInfo.getRemote(), "${branchName}:${branchName}");
+        try {
+          fetchResult.ourThrowExceptionIfFailed();
+        } catch (VcsException e) {
+          fetchResult.showNotificationIfFailed("Update of branch '${branchName}' failed");
+        }
+      }
+
+      @Override
+      public void onSuccess() {
+        VcsNotifier.getInstance(project).notifySuccess("Branch '${branchName}' updated");
+      }
+    }.queue();
+  }
+
+  protected String getRelationBaseDescription(SyncToRemoteStatus.Relation relation) {
+    String descriptionSpec = Match(relation).of(
+        Case($(SyncToRemoteStatus.Relation.Ahead), "ahead its remote"),
+        Case($(SyncToRemoteStatus.Relation.DivergedAndNewerThanRemote), "diverged (& newer) from its remote"),
+        Case($(SyncToRemoteStatus.Relation.DivergedAndOlderThanRemote), "diverged (& older) from its remote"),
+        Case($(SyncToRemoteStatus.Relation.InSync), "in sync to its remote"),
+        Case($(SyncToRemoteStatus.Relation.Untracked), "untracked"),
+        Case($(), "in unknown status '${relation.toString()}' to its remote"));
```

Comment:
This entire match is kinda repeated in `BasePushBranchAction`... maybe extract to some utils? Just the part the get `descriptionSpec`, not the there entire `...disabled because...` message